modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java

```diff
@@ -333,16 +325,142 @@ public static byte convertToByteExact(long x) {
+
+            if (rounded.compareTo((BigDecimal)INT_BOUNDS[4]) >= 0 && rounded.compareTo((BigDecimal)INT_BOUNDS[5]) <= 0)
+                return rounded;
+        }
+        else if (x instanceof Double) {
+            if (((Double)x).compareTo((Double)INT_BOUNDS[2]) < 0 && ((Double)x).compareTo((Double)INT_BOUNDS[3]) > 0)
+                return null;
+        }
+        else if (x instanceof Float) {
+            if (((Float)x).compareTo((Float)INT_BOUNDS[0]) < 0 && ((Float)x).compareTo((Float)INT_BOUNDS[1]) > 0)
+                return null;
+        }
+        else {
+            long longVal = x.longValue();
+
+            if (longVal >= Integer.MIN_VALUE && longVal <= Integer.MAX_VALUE)
+                return null;
+        }
+
+        throw new ArithmeticException(INTEGER.getName() + " overflow");
+    }
+
+    /** */
+    @Nullable private static BigDecimal checkShortBounds(Number x) {
+        if (x instanceof BigDecimal) {
+            BigDecimal rounded = ((BigDecimal)x).setScale(0, NUMERIC_ROUNDING_MODE);
+
+            if (rounded.compareTo((BigDecimal)SHORT_BOUNDS[4]) >= 0 && rounded.compareTo((BigDecimal)SHORT_BOUNDS[5]) <= 0)
+                return rounded;
+        }
+        else if (x instanceof Double) {
+            if (((Double)x).compareTo((Double)SHORT_BOUNDS[2]) < 0 && ((Double)x).compareTo((Double)SHORT_BOUNDS[3]) > 0)
+                return null;
+        }
+        else if (x instanceof Float) {
+            if (((Float)x).compareTo((Float)SHORT_BOUNDS[0]) < 0 && ((Float)x).compareTo((Float)SHORT_BOUNDS[1]) > 0)
+                return null;
+        }
+        else {
+            long longVal = x.longValue();
+
+            if (longVal >= Short.MIN_VALUE && longVal <= Short.MAX_VALUE)
+                return null;
+        }
+
+        throw new ArithmeticException(SMALLINT.getName() + " overflow");
+    }
+
+    /** */
+    @Nullable private static BigDecimal checkByteBounds(Number x) {
+        if (x instanceof BigDecimal) {
+            BigDecimal rounded = ((BigDecimal)x).setScale(0, NUMERIC_ROUNDING_MODE);
+
+            if (rounded.compareTo((BigDecimal)BYTE_BOUNDS[4]) >= 0 && rounded.compareTo((BigDecimal)BYTE_BOUNDS[5]) <= 0)
+                return rounded;
+        }
+        else if (x instanceof Double) {
+            if (((Double)x).compareTo((Double)BYTE_BOUNDS[2]) < 0 && ((Double)x).compareTo((Double)BYTE_BOUNDS[3]) > 0)
+                return null;
+        }
+        else if (x instanceof Float) {
+            if (((Float)x).compareTo((Float)BYTE_BOUNDS[0]) < 0 && ((Float)x).compareTo((Float)BYTE_BOUNDS[1]) > 0)
+                return null;
+        }
+        else {
+            long longVal = x.longValue();
+
+            if (longVal >= Byte.MIN_VALUE && longVal <= Byte.MAX_VALUE)
+                return null;
+        }
+
+        throw new ArithmeticException(TINYINT.getName() + " overflow");
+    }
+
+    /** */
+    private static double round(double x) {
```

Comment:
Perhaps `round` is not a good name. Becouse it prepares to round, but not do round itself.

Reply:
Didn't get it. it does round. Changes the value just like HALF_UP strategy.

Reply:
Method doesn't round value. It's adds or substracts 0.5 to the value (prepare to round). round method should return value without floating point.

Reply:
Renamed